numpy/_core/src/npysort/quicksort.hpp

```diff
@@ -0,0 +1,96 @@
+#ifndef NUMPY_SRC_COMMON_NPYSORT_QUICKSORT_HPP
+#define NUMPY_SRC_COMMON_NPYSORT_QUICKSORT_HPP
+
+#include "heapsort.hpp"
+#include "common.hpp"
+
+namespace np::sort {
+
+// pushing largest partition has upper bound of log2(n) space
+// we store two pointers each time
+constexpr size_t kQuickStack = sizeof(intptr_t) * 8 * 2;
+constexpr ptrdiff_t kQuickSmall = 15;
+
+// NUMERIC SORTS
+template <typename T>
+inline void Quick(T *start, SSize num)
```

Comment:
>  . Could we reuse this implementation in quicksort.cpp?

No, we need to move into a separate header which I did but with some changes as I mentioned within commit log:

> Refactored the C++ implementation of quicksort into a separate header.
  This is to ensure its availability as a fallback when SIMD kernel versions of quicksort are not available.

> The refactored quicksort is now encapsulated within the `np::sort` namespace, and is no longer relying on tags.
  Future patches will transition the rest of the sorting functions to utilize C++ operators instead of relying on tags.

> This patch also introduced a new C++ alias `np::SSize`, mapping to `Py_ssize_t`, and
  implemented a new utility function `np::BitScanReverse` for efficiently performing floor(log2(x)) operations,
  offering improved performance over the existing C function `npy_get_msb()`.